pkg/roll/run_update.go

```diff
@@ -0,0 +1,305 @@
+				u.logger.WithFields(logrus.Fields{
+					"old": oldNodes,
+					"new": newNodes,
+				}).Debugln("Upgrade complete")
+				break ROLL_LOOP
+			}
+			next := algorithm(oldNodes.desired, newNodes.desired, u.DesiredReplicas, u.MinimumReplicas)
+			u.logger.WithFields(logrus.Fields{
+				"old":  oldNodes,
+				"new":  newNodes,
+				"next": next,
+			}).Infoln("Undergoing next update")
+
+			if next > 0 {
+				if u.DeletePods {
+					u.rcs.AddDesiredReplicas(u.OldRC, -next)
+				}
+				u.rcs.AddDesiredReplicas(u.NewRC, next)
+			}
+		}
+	}
+
+	// rollout complete, clean up old RC
+	u.logger.NoFields().Infoln("Cleaning up old RC")
+	if err := u.rcs.SetDesiredReplicas(u.OldRC, 0); err != nil {
+		return err
+	}
+	if err := u.rcs.Enable(u.OldRC); err != nil {
+		return err
+	}
+	if err := u.rcs.Delete(u.OldRC, false); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u update) lock(session string) error {
+	if success, err := u.rcs.LockWrite(u.NewRC, session); err != nil {
+		return err
+	} else if !success {
+		return fmt.Errorf("could not lock new %s", u.NewRC)
+	}
+
+	if success, err := u.rcs.LockWrite(u.OldRC, session); err != nil {
+		return err
+	} else if !success {
+		return fmt.Errorf("could not lock old %s", u.OldRC)
+	}
+
+	return nil
+}
+
+func (u update) enable() error {
+	err := u.rcs.Enable(u.NewRC)
+	if err != nil {
+		return err
+	}
+
+	if !u.DeletePods {
+		err = u.rcs.Disable(u.OldRC)
+	} else {
+		err = u.rcs.Enable(u.OldRC)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type rcNodeCounts struct {
+	desired int
+	current int
+	healthy int
+}
+
+func (u update) countHealthy(id rcf.ID, checks map[string]health.Result) (rcNodeCounts, error) {
+	ret := rcNodeCounts{}
+	rcFields, err := u.rcs.Get(id)
+	if err != nil {
+		return ret, err
+	}
+	ret.desired = rcFields.ReplicasDesired
+
+	nodes, err := rc.New(rcFields, u.kps, u.rcs, u.sched, u.labeler, u.logger).CurrentNodes()
+	if err != nil {
+		return ret, err
+	}
+	ret.current = len(nodes)
+
+	for _, node := range nodes {
+		if hres, ok := checks[node]; ok && hres.Status == health.Passing {
+			ret.healthy++
+		}
+	}
+	return ret, err
+}
+
+// the roll algorithm defines how to mutate RCs over time. it takes four args:
+// - old: the number of nodes on the old RC
+// - new: the number of nodes on the new RC
+// - want: the number of nodes desired on the new RC (ie the target)
+// - need: the number of nodes that must always be up (ie the minimum)
+// given these four arguments, rollAlgorithm returns the number of nodes to add
+// to the new RC (and, if DeletePods=true, to delete from the old)
+//
+// returns zero under the following circumstances:
+// - new >= want (the update is done)
+// - old+new <= need (at or below the minimum, update has to block)
+func algorithm(old, new, want, need int) int {
+	// how much "headroom" do we have between the number of nodes that are
+	// currently healthy, and the number that must be healthy?
```

Comment:
this comment is confusing to me. don't `old` and `new` just signify the number desired rather than the number that are healthy